services/auth/httpsign.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package auth
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/web/middleware"
+
+	"github.com/go-fed/httpsig"
+	"golang.org/x/crypto/ssh"
+)
+
+// Ensure the struct implements the interface.
+var (
+	_ Method = &HTTPSign{}
+	_ Named  = &HTTPSign{}
+)
+
+// HTTPSign implements the Auth interface and authenticates requests (API requests
+// only) by looking for http signature data in the "Signature" header.
+// more information can be found on https://github.com/go-fed/httpsig
+type HTTPSign struct{}
+
+// Name represents the name of auth method
+func (h *HTTPSign) Name() string {
+	return "httpsign"
+}
+
+// Verify extracts and validates HTTPsign from the Signature header of the request and returns
+// the corresponding user object on successful validation.
+// Returns nil if header is empty or validation fails.
+func (h *HTTPSign) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *user_model.User {
+	// HTTPSign authentication should only fire on API
+	if !middleware.IsAPIPath(req) {
+		return nil
+	}
+
+	sigHead := req.Header.Get("Signature")
+	if len(sigHead) == 0 {
+		return nil
+	}
+
+	var (
+		u       *user_model.User
+		validpk *asymkey_model.PublicKey
+		err     error
+	)
+
+	// Handle SSH certificates
+	if len(req.Header.Get("x-ssh-certificate")) != 0 {
```

Comment:
Generally better to put the header keys in Canonical form.

```suggestion
	if len(req.Header.Get("X-Ssh-Certificate")) != 0 {
```